torch/storage.py

```diff
@@ -163,3 +139,288 @@ def _load_from_bytes(b):
+        torch.half: 'HalfStorage',
+        torch.long: 'LongStorage',
+        torch.int: 'IntStorage',
+        torch.int16: 'ShortStorage',
+        torch.int8: 'CharStorage',
+        torch.uint8: 'ByteStorage',
+        torch.bool: 'BoolStorage',
+        torch.bfloat16: 'BFloat16Storage',
+        torch.cdouble: 'ComplexDoubleStorage',
+        torch.cfloat: 'ComplexFloatStorage',
+        torch.qint8: 'QInt8Storage',
+        torch.qint32: 'QInt32Storage',
+        torch.quint8: 'QUInt8Storage',
+        torch.quint4x2: 'QUInt4x2Storage',
+    }
+
+def _pickle_storage_type_to_dtype_map():
+    if not hasattr(_pickle_storage_type_to_dtype_map, 'cache'):
+        _pickle_storage_type_to_dtype_map.cache = {  # type: ignore[attr-defined]
+            val: key for key, val in _dtype_to_pickle_storage_type_map().items()}
+    return _pickle_storage_type_to_dtype_map.cache  # type: ignore[attr-defined]
+
+class TypedStorage(_StorageOverrides):
+    is_sparse = False
+
+    @property
+    def is_cuda(self):
+        return self.storage.is_cuda
+
+    def fill_(self, value):
+        self[0:len(self)] = value
+        return self
+
+    def __init__(self, *args, **kwargs):
+        if 'wrap_storage' in kwargs:
+            assert len(args) == 0, (
+                "No positional arguments should be given when using "
+                "'wrap_storage'")
+
+            if type(self) == TypedStorage:
+                assert 'dtype' in kwargs, (
+                    "When using 'wrap_storage', 'dtype' also must be specified")
+                assert len(kwargs) == 2, (
+                    "Only 'wrap_storage' and 'dtype' should be given, but got: "
+                    f"{kwargs}")
+                dtype = kwargs['dtype']
+                assert isinstance(dtype, torch.dtype)
+                self.dtype = dtype
+
+            else:
+                assert hasattr(self, 'dtype')
+                assert len(kwargs) == 1, (
+                    f"Only 'wrap_storage' should be given, but got: {kwargs.keys()}")
+                dtype = self.dtype
+
+            storage = kwargs['wrap_storage']
+
+            # TODO: change `self.storage` to `self._storage`
+            if isinstance(storage, TypedStorage):
+                # TODO: Consider whether we should actually allow this or not.
+                # I think it might be useful for .float(), .double(), etc
+                # methods?
+                self.storage = storage.storage
+
+            else:
+                assert isinstance(storage, eval(storage.__module__ + ".ByteStorage"))
+                self.storage = storage
+        else:
+            assert type(self) != TypedStorage, (
+                "Calling __init__ this way is only supported in TypedStorage's "
+                "child classes. TypedStorage can only be directly instantiated "
+                "when kwargs 'wrap_storage' and 'dtype' are given.")
+
+            # TODO: Still need to override some of the constructor behavior.
```

Comment:
Note: This constructor will have to get fairly complex to account for all the cases where we need to override `ByteStorage`'s constructor. I don't know if there's any simpler approach

Reply:
(Yes this is a layer violation but it is very mild because this is a legacy BC layer and we gonna delete it eventually anyway)